File: outreach/outreach/report/student_attendance/student_attendance.py
import frappe

ATTENDED_STATUSES = ("Attended", "Partially Attended")


def execute(filters=None):
    filters = frappe._dict(filters or {})

    stacks = filters.get("session_stacks") or []
    if isinstance(stacks, str):
        stacks = [s.strip() for s in stacks.split(",") if s.strip()]

    # Base columns
    base_columns = [
        {
            "fieldname": "student",
            "label": "Student",
            "fieldtype": "Link",
            "options": "Student",
            "width": 180,
        },
        {
            "fieldname": "mentor",
            "label": "Mentor",
            "fieldtype": "Link",
            "options": "User",
            "width": 180,
        },
    ]

    if not stacks:
        return base_columns, []

    # 1) Students (mentor comes from Student)
    students = frappe.get_all(
        "Student",
        fields=["name", "mentor"],
        order_by="name asc",
    )

    # 2) Dynamic stack columns
    columns = list(base_columns)
    for stack in stacks:
        columns.append({
            "fieldname": f"stack_{frappe.scrub(stack)}",
            "label": stack,
            "fieldtype": "Data",
            "width": 70,
        })

    # 3) Followup Sessions → stack mapping
    sessions = frappe.get_all(
        "Followup Session",
        filters={"session_stack": ["in", stacks]},
        fields=["name", "session_stack"],
    )
    if not sessions:
        return columns, [
            {"student": s["name"], "mentor": s["mentor"]}
            for s in students
        ]

    session_stack_map = {s["name"]: s["session_stack"] for s in sessions}
    session_names = list(session_stack_map.keys())

    # 4) Attendance from Followup Record
    attendance_rows = frappe.get_all(
        "Followup Record",
        filters={
            "followup_session": ["in", session_names],
            "call_status": ["in", ATTENDED_STATUSES],
        },
        fields=["student", "followup_session"],
    )

    present = set()
    for row in attendance_rows:
        student = row.get("student")
        session = row.get("followup_session")
        if student and session in session_stack_map:
            present.add((student, session_stack_map[session]))

    # 5) Build data
    data = []
    for s in students:
        row = {
            "student": s["name"],
            "mentor": s.get("mentor"),
        }
        for stack in stacks:
            key = (s["name"], stack)
            row[f"stack_{frappe.scrub(stack)}"] = "P" if key in present else ""
        data.append(row)

    return columns, data

File: outreach/outreach/report/student_attendance/student_attendance.js
frappe.query_reports["Student Attendance"] = {
    filters: [
        {
            fieldname: "session_stacks",
            label: __("Session Stacks"),
            fieldtype: "MultiSelectList",
            options: "Session Stack",
            reqd: 1,
            get_data: function (txt) {
                return frappe.db.get_link_options("Session Stack", txt);
            },
        },
    ],

    onload: function (report) {
        report.page.add_inner_button(__("Assign Followup"), function () {
            // IMPORTANT: read rows at click-time, not onload
            const rows = (typeof report.get_data === "function")
                ? report.get_data()
                : (report.data || []);

            const student_names = Array.from(
                new Set(
                    (rows || [])
                        .map((row) => row.student)
                        .filter((s) => !!s)
                )
            );

            if (!student_names.length) {
                frappe.msgprint(__("No students in current result to assign."));
                return;
            }

            // Dialog: same UX as Student List "Assign to Followup"
            let d = new frappe.ui.Dialog({
                title: __("Assign These Students to Followup"),
                fields: [
                    {
                        label: "Followup Session",
                        fieldname: "followup_session",
                        fieldtype: "Select",
                        reqd: 1,
                        options: [],
                        onchange: function () {
                            const selected_session = d.get_value("followup_session");
                            const session_data = d.sessionMap[selected_session];
                            if (session_data) {
                                d.set_value("session_stack", session_data.session_stack);
                                d.set_value("deadline", session_data.deadline);
                            }
                        },
                    },
                    {
                        label: "Session Stack",
                        fieldname: "session_stack",
                        fieldtype: "Data",
                        read_only: 1,
                    },
                    {
                        label: "Deadline",
                        fieldname: "deadline",
                        fieldtype: "Datetime",
                        read_only: 1,
                    },
                    {
                        label: "Caller",
                        fieldname: "caller",
                        fieldtype: "Link",
                        options: "User",
                        reqd: 1,
                    },
                ],
                primary_action_label: __("Assign"),
                primary_action(values) {
                    if (!values.followup_session || !values.caller) {
                        frappe.msgprint(__("Followup Session and Caller are required."));
                        return;
                    }

                    frappe.call({
                        method: "outreach.api.followup_assignment.assign_students_to_followup",
                        args: {
                            students: student_names,
                            followup_session: values.followup_session,
                            caller: values.caller,
                        },
                        freeze: true,
                        freeze_message: __("Creating Followup Records..."),
                        callback: (r) => {
                            if (!r.exc) {
                                frappe.msgprint(
                                    __(
                                        "Created {0} Followup Records successfully.",
                                        [r.message && r.message.count || student_names.length]
                                    )
                                );
                                d.hide();
                            }
                        },
                    });
                },
            });

            // Load active followup sessions (reuse same API as Student List)
            frappe.call({
                method: "outreach.api.followup_assignment.get_active_followup_sessions",
                callback: (r) => {
                    const sessions = r.message || [];
                    d.sessionMap = {};
                    d.set_df_property(
                        "followup_session",
                        "options",
                        sessions.map((s) => s.label).join("\n")
                    );
                    sessions.forEach((s) => {
                        d.sessionMap[s.label] = s;
                    });
                },
            });

            d.show();
        });
    },
};